packages/app/src/components/DndContext.js

```diff
@@ -0,0 +1,143 @@
+import React, { Component } from 'react'
+import { connect } from 'react-redux'
+import PropTypes from 'prop-types'
+import { DragDropContext } from 'react-beautiful-dnd'
+import { canDimensionBeAddedToAxis } from '@dhis2/analytics'
+
+import * as fromReducers from '../reducers'
+import * as fromActions from '../actions'
+import { getAdaptedUiByType } from '../modules/ui'
+import { SOURCE_DIMENSIONS } from '../modules/layout'
+
+class DndContext extends Component {
+    rearrangeLayoutDimensions = ({
+        sourceAxisId,
+        sourceIndex,
+        destinationAxisId,
+        destinationIndex,
+    }) => {
+        const layout = this.props.layout
+
+        const sourceList = Array.from(layout[sourceAxisId])
+        const [moved] = sourceList.splice(sourceIndex, 1)
+
+        if (sourceAxisId === destinationAxisId) {
+            sourceList.splice(destinationIndex, 0, moved)
+
+            this.props.onReorderDimensions({
+                ...layout,
+                [sourceAxisId]: sourceList,
+            })
+        } else {
+            if (
+                canDimensionBeAddedToAxis(
+                    this.props.type,
+                    layout[destinationAxisId],
+                    destinationAxisId
+                )
+            ) {
+                this.props.onAddDimensions({
+                    [moved]: destinationAxisId,
+                })
+            }
+        }
+    }
+
+    addDimensionToLayout = ({ axisId, index, dimensionId }) => {
+        const { layout, type } = this.props
+
+        if (!canDimensionBeAddedToAxis(type, layout[axisId], axisId)) {
+            return
+        }
+
+        const destList = Array.from(layout[axisId])
+
+        destList.splice(index, 0, dimensionId)
+        const reorderedDimensions = { [axisId]: destList }
+
+        this.props.onReorderDimensions({
```

Comment:
IIUC the case here is a dimension dragged from the DimensionPanel to the layout.
I believe this should use the `onAddDimensions` which runs the layout rules validation.